lib/matplotlib/testing/_conversion_cache.py

```diff
@@ -0,0 +1,211 @@
+                        unicode_literals)
+
+import hashlib
+import shutil
+import os
+import warnings
+
+from matplotlib import _get_cachedir
+from matplotlib import cbook
+from matplotlib import checkdep_ghostscript
+from matplotlib import checkdep_inkscape
+
+
+class _ConversionCache(object):
+    """A cache that stores png files converted from svg or pdf formats.
+
+    The image comparison test cases compare svg and pdf files by
+    converting them to png files. When a test case has produced a
+    file, e.g. result.pdf, it queries this cache by the pathname
+    '/path/to/result_images/result.pdf'. The cache computes a hash of
+    the file (and the version of the external software used to convert
+    the file) and if a result by that hash value is available, writes
+    the data to the output location supplied by the caller. Otherwise
+    the test case has to run the conversion and can then insert the
+    result into the cache.
+
+    Parameters
+    ----------
+    directory : str, optional
+        Files are stored in this directory. Defaults to `'test_cache'` in
+        the overall Matplotlib cache directory.
+    max_size : int, optional
+        The flush method will delete files until their combined size is
+        under this limit, in bytes. Defaults to 100 megabytes.
+
+    """
+
+    def __init__(self, directory=None, max_size=int(1e8)):
+        self.gets = set()
+        self.hits = set()
+        if directory is not None:
+            self.cachedir = directory
+        else:
+            self.cachedir = self.get_cache_dir()
+        self.ensure_cache_dir()
+        if not isinstance(max_size, int):
+            raise ValueError("max_size is of type %s, expected int" %
+                             type(max_size))
+        self.max_size = max_size
+        self.cached_ext = '.png'
+        self.converter_version = {}
+        try:
+            self.converter_version['.pdf'] = \
+                checkdep_ghostscript()[1].encode('utf-8')
+        except:
+            pass
+        try:
+            self.converter_version['.svg'] = \
+                checkdep_inkscape().encode('utf-8')
+        except:
+            pass
+        self.hash_cache = {}
+
+    def get(self, filename, newname):
+        """Query the cache.
+
+        Parameters
+        ----------
+        filename : str
+            Full path to the original file.
+        newname : str
+            Path to which the result should be written.
+
+        Returns
+        -------
+        bool
+            True if the file was found in the cache and is now written
+            to `newname`.
+        """
+        self.gets.add(filename)
+        hash_value = self._get_file_hash(filename)
+        cached_file = os.path.join(self.cachedir, hash_value + self.cached_ext)
+        with cbook.Locked(self.cachedir):
+            if os.path.exists(cached_file):
+                shutil.copyfile(cached_file, newname)
```

Comment:
Should we {sym,hard}link instead of copying?
(On Windows, admin rights are needed to symlink but not to hardlink.  Python only added Windows support for either in Py3.2 (https://docs.python.org/3.6/library/os.html#os.link, https://docs.python.org/3.6/library/os.html#os.symlink).